Add early template validation

kmarinushkin opened the pull request:
This PR is a proposed fix for an issue https://github.com/os-climate/ITR/issues/119

kmarinushkin commented:
Hello @dp90, @MichaelTiemannOSC,

in this PR, i tried to resolve on of open issues in the repo.
Please feel free to correct me, if I didn't understand anything in the issue or the code.
I'm looking forward for your feedback

MichaelTiemannOSC commented:
Would it be better to use .any() instead of .max()?  any() can return the first null value, whereas max() must scan all null values.

kmarinushkin commented:
Interesting proposal, @MichaelTiemannOSC! I didn't think about this detail. Maybe I even could replace a chain of `is_null().max()` with a single `any()`. I will make a change later today

kmarinushkin commented on ITR/data/template.py:
```diff
@@ -125,6 +125,13 @@ def _fixup_name(x):
         df_fundamentals = df_fundamentals[df_fundamentals.columns.difference(historic_columns, sort=False)]
 
+        # Checking missing company ids
+        missing_company_ids = df_fundamentals[ColumnsConfig.COMPANY_ID].isnull().any()
```
`.any()` it is!

kmarinushkin commented:
Hello @dp90, 

you once opened an issue in this repo (https://github.com/os-climate/ITR/issues/119)
In an attempt to be helpful, i opened this PR, as a proposal, to resolve the issue.

May i ask you, to review my changes, and let me know, if they satisfy your vision of the issue

dp90 commented:
Hi @kmarinushkin ,

Thanks for the contribution! Regarding your question about whether the PR tackles the issue I opened previously, I'm not sure that it does. Although the checks you added by themselves make perfect sense, the issue I opened is about which part of the code performs validation of the data and at what point. 
The original setup depends on the inherited Pydantic validation in the ICompanyData object (see the checks and validations in the definition of ICompanyData in interfaces.py). In that setup, the data - potentially coming from different types of sources, such as Excel files, JSON, etc., - should first be cast to ICompanyData instances (that have some attributes that are themselves classes inheriting from the Pydantic model with their own validation), so that validation is performed right away. 

However, the TemplateProviderCompany class, before casting the data to the ICompanyData objects (and its attributes), performs some conversions that depend on some data being present/correct. As mentioned in the issue, this can lead to errors that are difficult to trace. So we'd need validation of the data before we can cast it to the object types that already perform the data validation. 

In short, the point of the issue was that - even though the check you added might be the final check needed to make sure the conversions always work at this moment - I think we should avoid the double validation. A solution could be to try to cast the data to the Pydantic-validated objects earlier, although I'm afraid I don't have the resources to check the feasibility of that option at the moment myself.

kmarinushkin commented:
@dp90 all right, thank you for clarification, David!

this PR was my first attempt to bring value to the ITR project, before i managed to contact Michael.
in the recent weeks, we discussed with him the roadmap for work, where my contribution might be helpful in the current state of the tool. We prioritised the "scopes" work, so, i don't plan to continue with data validation for now.

i want to close this PR - i don't like hanging unfinished work. We may merge it, if you think it does something useful. Or we may close it w/o merging, since it doesn't resolve the issue it intended to.

Which way would you prefer us to follow?

dp90 commented:
Let's merge it, as it indeed adds something useful, and keep the issue open. 
The final step would be to increment the version number in setup.py (unfortunately still a manual step) that should also fix the final failing check.

kmarinushkin commented:
Thank you for the guidance, David!

i think that changes in this PR are too minor, to cause a version increment.
I think the version increment became necessary, because i opened a PR to `main` branch.
So, i changed the target branch to `develop`, and rebased.
This should fit better, if i understand branching in ITR correctly

kmarinushkin commented:
Hello @MichaelTiemannOSC, I have an approval in this PR, but I don't have write permission to the repo, to merge.
May I ask you to click the button?
